core/src/com/google/inject/internal/aop/EnhancerBuilderImpl.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2019 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.inject.internal.aop;
+
+import static com.google.inject.internal.aop.ClassBuilding.signature;
+import static com.google.inject.internal.aop.ClassBuilding.visitMembers;
+import static com.google.inject.internal.aop.ClassDefining.hasPackageAccess;
+import static java.lang.reflect.Modifier.FINAL;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.inject.internal.BytecodeGen;
+import java.lang.reflect.Executable;
+import java.lang.reflect.Method;
+import java.util.BitSet;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+/**
+ * Builder of enhancers that provide method interception via bytecode generation.
+ *
+ * @author mcculls@gmail.com (Stuart McCulloch)
+ */
+final class EnhancerBuilderImpl implements BytecodeGen.EnhancerBuilder {
+
+  /** Lazy association between classes and their generated enhancers. */
+  private static final ClassValue<Map<BitSet, Function<String, BiFunction>>> ENHANCERS =
+      new ClassValue<Map<BitSet, Function<String, BiFunction>>>() {
+        @Override
+        protected Map<BitSet, Function<String, BiFunction>> computeValue(Class<?> hostClass) {
+          return new HashMap<>();
+        }
+      };
+
+  private final Class<?> hostClass;
+
+  private final Method[] enhanceableMethods;
+
+  private final Map<Method, Method> bridgeDelegates;
+
+  public EnhancerBuilderImpl(
+      Class<?> hostClass,
+      Collection<Method> enhanceableMethods,
+      Map<Method, Method> bridgeDelegates) {
+
+    this.hostClass = hostClass;
+    this.enhanceableMethods = enhanceableMethods.toArray(new Method[enhanceableMethods.size()]);
+    this.bridgeDelegates = ImmutableMap.copyOf(bridgeDelegates);
+  }
+
+  @Override
+  public Method[] getEnhanceableMethods() {
+    return enhanceableMethods;
+  }
+
+  @Override
+  public Function<String, BiFunction> buildEnhancer(BitSet methodIndices) {
+    if ((hostClass.getModifiers() & FINAL) != 0) {
+      throw new IllegalArgumentException("Cannot subclass final " + hostClass);
+    }
```

Comment:
This check can't be in the constructor because that happens before we match enhanceable methods against the list of potential interceptors. If we did the check in the constructor then we would end up throwing exceptions for final classes that happened to match an interceptor at the class-level, but didn't match it the method-level (ie. there weren't any matching enhanceable methods.)

In other words we need to defer the check until we know we definitely want to enhance the class, as opposed to when we might want to enhance it.

Reply:
That makes sense. Thanks!